[Feature] Manual inline feedback for Programming Exercises and fix repo is locked label

fde312 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
At the moment tutors or instructors are not able to give feedback for a specific line of code for programming exercises.

### Description
<!-- Describe your changes in detail -->
- Use CodeEditorAceComponent for the inline feedbacks (referenced feedback)
- Created InlineFeedbackComponent which is displayed in the AceEditor
- Allow tutors to add unreferenced and general feedback 
- Add label which indicates how many points the assessment has
- UI fix for 'repository is locked' label in CodeEditorStudentComponent

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise and participate with several different users
4. Login as tutor/instructor and assess the programming exercise
5. When hovering over the gutter (line numbers on the left) an plus icon appears and click
6. Add inline feedback
7. Add unreferenced feedback
8. Add general feedback
9. Check if the points label adapts accordingly
10. Submit your assessment and see if everything is updated accordingly and works as expected (check also student view)
11. Check also for complaints and exam

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
* code-editor-tutor-assessment-inline-feedback.component.ts: 95%
* code-editor-tutor-assessment-container.component.ts: 69.49%
* code-editor-ace.component.ts: 66.06%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
As a Tutor:

Hover over code and click on plus button..
![image](https://user-images.githubusercontent.com/41108487/93323075-6b36a880-f814-11ea-87ef-396f60087975.png)

.. inline comment appears and you can enter feedback text and points. Color changes depending on the given points..
![image](https://user-images.githubusercontent.com/41108487/93323161-89040d80-f814-11ea-9359-8b1e0ab54471.png)
..after clicking on save.
![image](https://user-images.githubusercontent.com/41108487/93323190-93260c00-f814-11ea-9540-f51af36de6ad.png)

Unrefenreced and general feedback as known from the other exercises:
![image](https://user-images.githubusercontent.com/41108487/93324073-b1d8d280-f815-11ea-9871-ee09d0a80604.png)



Student view:

<img width="1370" alt="image" src="https://user-images.githubusercontent.com/41108487/93323588-19425280-f815-11ea-850d-cdded18221aa.png">

Feedback info:
<img width="906" alt="image" src="https://user-images.githubusercontent.com/41108487/93323717-4131b600-f815-11ea-9b5e-7728a31b089d.png">

JonasPetry commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-inline-feedback.component.ts:
```diff
@@ -0,0 +1,83 @@
+        return this._feedback;
+    }
+    set feedback(feedback: Feedback) {
+        this._feedback = feedback || new Feedback();
```
I don't think that we use the _ notation so far

fde312 replied:
see comment above :) I just checked and we have it the same way e.g. for drag and drop and short answer questions

JonasPetry commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-inline-feedback.component.ts:
```diff
@@ -0,0 +1,83 @@
+    @Input()
+    get feedback(): Feedback {
+        return this._feedback;
+    }
```
Does this need to be a getter? I think this can be a standard input 🤔

fde312 replied:
I wanted to intercept the input with a setter and this is how it is written down in the angular docu:
https://angular.io/guide/component-interaction#intercept-input-property-changes-with-a-setter
Just used it that way.

schultek reviewed (approved):
Everything works great and the code looks good. 👍
- Giving inline feedback works great across lines and different files.
- The manual score is calculated correctly from all feedbacks (general, line, unreferenced)
- The manual result is saved and submitted correctly and can be updated.
- The feedback is displayed correctly to the student and inline in the code editor.

I think this is a very useful feature. Great work. 🚀

One small thing. When the manual result is set, the tasks in the exercise instructions all display 0 of x tests passed. It of course would be great to still have the test case feedback here, but i think this is something for a follow-up PR.

stefanwaldhauser reviewed (approved):
Everything works as expected! Great feature and great work @fde312  :)

kloessst reviewed (changes requested):
Really nice feature. Works mostly fine. I just noticed two things:

After I complained about my manual assessment, the complaint was instantly shown to be addressed in the bar diagram, but not in the complaints list. What is the expected behavior here?

![complaint](https://user-images.githubusercontent.com/16407766/94293730-08a18300-ff5f-11ea-8f1d-8569cae9adac.png)
![complaintList](https://user-images.githubusercontent.com/16407766/94293733-093a1980-ff5f-11ea-92da-123078a31aca.png)

Afterwards I tried to override the assessment with the complaint but couldn't, seems like the server is throwing an exception:

![errorOverview](https://user-images.githubusercontent.com/16407766/94294012-7352be80-ff5f-11ea-99d4-f7d464f0d6cd.png)
![errorDetails](https://user-images.githubusercontent.com/16407766/94294008-72ba2800-ff5f-11ea-8b95-160b54587fcb.png)

Can you reproduce this?

anditurdiu commented:
> Really nice feature. Works mostly fine. I just noticed two things:
> 
> After I complained about my manual assessment, the complaint was instantly shown to be addressed in the bar diagram, but not in the complaints list. What is the expected behavior here?
> 
> ![complaint](https://user-images.githubusercontent.com/16407766/94293730-08a18300-ff5f-11ea-8f1d-8569cae9adac.png)
> ![complaintList](https://user-images.githubusercontent.com/16407766/94293733-093a1980-ff5f-11ea-92da-123078a31aca.png)
> 
> Afterwards I tried to override the assessment with the complaint but couldn't, seems like the server is throwing an exception:
> 
> ![errorOverview](https://user-images.githubusercontent.com/16407766/94294012-7352be80-ff5f-11ea-99d4-f7d464f0d6cd.png)
> ![errorDetails](https://user-images.githubusercontent.com/16407766/94294008-72ba2800-ff5f-11ea-8b95-160b54587fcb.png)
> 
> Can you reproduce this?

I believe it is expected behaviour that you cannot see the complaints about your own assessment. However, you see in the status that there are complaints

fde312 commented:
> > Really nice feature. Works mostly fine. I just noticed two things:
> > After I complained about my manual assessment, the complaint was instantly shown to be addressed in the bar diagram, but not in the complaints list. What is the expected behavior here?
> > ![complaint](https://user-images.githubusercontent.com/16407766/94293730-08a18300-ff5f-11ea-8f1d-8569cae9adac.png)
> > ![complaintList](https://user-images.githubusercontent.com/16407766/94293733-093a1980-ff5f-11ea-92da-123078a31aca.png)
> > Afterwards I tried to override the assessment with the complaint but couldn't, seems like the server is throwing an exception:
> > ![errorOverview](https://user-images.githubusercontent.com/16407766/94294012-7352be80-ff5f-11ea-99d4-f7d464f0d6cd.png)
> > ![errorDetails](https://user-images.githubusercontent.com/16407766/94294008-72ba2800-ff5f-11ea-8b95-160b54587fcb.png)
> > Can you reproduce this?
> 
> I believe it is expected behaviour that you cannot see the complaints about your own assessment. However, you see in the status that there are complaints

@kloessst I could reproduce your error and fixed it. I found also that the result was not updated correctly after submitting the complaint. I fixed that too. Now everything should work as expected :) 

Related to your first point I created an issue. @anditurdiu is right, you should not be able to see the complaint about your own assessment, but it should say 0/1 instead of 1/1 (see #2124)

MaisiKoleni reviewed (approved):
Works really well, apart from the issue that is now being discussed in #2129.

The creation of multiple inline comments for the same line of code is fixed now.

I only found a very minor layout problem with half a line of  whitespace below the inline comments that is barely noticeable and does not affect usability. In a future update it would be good to move the complaint functionality directly to the code editor.

Good work!

kloessst reviewed (dismissed):
Great feature. Works really well. Code also looks good to me. Well done!

balazs-czopf commented on src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.scss:
```diff
@@ -12,6 +12,38 @@ Ace Editor
+
+            fa-icon:hover {
+                font-size: 15px;
+                transition: font-size 0.35s ease;
```
I think the animation for the icon getting larger could be a bit faster so I would suggest 200ms here, but this is really just personal preference 😄

fde312 replied:
If anyone else is feeling the same, we can change it :)

balazs-czopf commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-inline-feedback.component.html:
```diff
@@ -0,0 +1,53 @@
+<div
+    class="border rounded"
+    [style.max-width]="'780px'"
```
Do we really need that here? If there is no special reason I would omit that so that instead of getting this: 
![image](https://user-images.githubusercontent.com/63286106/94397016-5555ab80-0163-11eb-8e03-80f0e2fd1b7c.png)
we could have that div simply adjust its with to the parent like this:
![image](https://user-images.githubusercontent.com/63286106/94397138-7a4a1e80-0163-11eb-89c6-67f797170376.png)

fde312 replied:
I oriented this how it is done for GitHub. There the comment does not have the full width either and is around 780px width :) 
![image](https://user-images.githubusercontent.com/41108487/94404022-999a7900-016e-11eb-9fb3-52d526f04da6.png)

balazs-czopf replied:
oh okay I see :)

balazs-czopf reviewed (dismissed):
I left two inline comments regarding some design issues but other than that this feature is really awesome 🚀 nice work!

anditurdiu reviewed (approved):
Code looks good and the feature works as expected! Good Job!

JonasPetry reviewed (dismissed):
Good job 👍

filip-gregurevic reviewed (approved):
Nice work! 🚀

krusche commented on src/main/webapp/app/entities/feedback.model.ts:
```diff
@@ -19,6 +19,7 @@ export enum FeedbackType {
 }
 
 export const STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER = 'SCAFeedbackIdentifier:';
+export const MANUAL_ASSESSMENT_IDENTIFIER = 'MAFeedbackIdentifier:';
```
do we really need this? We have FeedbackType.MANUAL for this

fde312 replied:
I removed it :)

krusche reviewed (changes requested):
Why do we need MANUAL_ASSESSMENT_IDENTIFIER?
We have FeedbackType.Manual for this
Can you please remove it?

Also in the feedback popup the red text states "Feedback" twice, which is not really necessary

krusche commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-inline-feedback.component.ts:
```diff
@@ -0,0 +1,83 @@
+     */
+    updateFeedback() {
+        this.feedback.type = this.MANUAL;
+        this.feedback.reference = `${MANUAL_ASSESSMENT_IDENTIFIER}_file:${this.fileName}_line:${this.codeLine}`;
```
For the future, it would be nice, if tutors can also select certain characters of a line. I think this could be extended using `file:${this.fileName}_line:${this.codeLine}_characters:{start}:{end}`

krusche replied:
It would be more important that `fileName` includes the path, in the (unlikely) case that there are two files with the same name

fde312 replied:
The reference contains the whole path of the file. I renamed the variable to have a consistent name throughout the components. :)

derLalla commented on src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts:
```diff
@@ -129,6 +161,22 @@ export class CodeEditorAceComponent implements AfterViewInit, OnChanges, OnDestr
+                this.fileFeedbacks = this.feedbacks.filter((feedback) => feedback.reference && feedback.reference.includes(this.selectedFile));
+                this.fileFeedbackPerLine = {};
+                this.fileFeedbacks.forEach((feedback) => {
+                    const line: number = +feedback.reference!.split('line:')[1];
```
is the + in `+feedback.reference!.split('line:')[1];` a typo?

fde312 replied:
Nope it is not, with that you transform a `string` to a `number` :)

derLalla commented on src/test/javascript/spec/component/programming-assessment/programming-assessment-inline-feedback.component.spec.ts:
```diff
@@ -0,0 +1,75 @@
+                };
+                fixture = TestBed.createComponent(CodeEditorTutorAssessmentInlineFeedbackComponent);
+                comp = fixture.componentInstance;
+                debugElement = fixture.debugElement;
```
`debugElement` should be removed as stated from GitHub Actions

![image](https://user-images.githubusercontent.com/33484318/94557940-8a472880-025f-11eb-978e-05e4d11cbfb8.png)

fde312 replied:
done :)

fde312 commented:
> Why do we need MANUAL_ASSESSMENT_IDENTIFIER?
> We have FeedbackType.Manual for this
> Can you please remove it?
> 
> Also in the feedback popup the red text states "Feedback" twice, which is not really necessary

I removed the MANUAL_ASSESSMENT_IDENTIFIER and just use feedback.Type === FeedbackType.MANUAL as check.
The text of the feedback is also changed, see screenshot below:
![image](https://user-images.githubusercontent.com/41108487/94558192-ee69ec80-025f-11eb-90e0-e0557f19290e.png)

stefanwaldhauser reviewed (approved):
reapprove

kloessst reviewed (approved):
Re-approve

jpbernius commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-inline-feedback.component.ts:
```diff
@@ -0,0 +1,83 @@
+})
+export class CodeEditorTutorAssessmentInlineFeedbackComponent {
+    MANUAL = FeedbackType.MANUAL;
+    @Input()
```
So apparently this does work, but `@Input` on the `set`ter might make more sense?
